pkg/providers/ibmcloud.go

```diff
@@ -0,0 +1,331 @@
+	rand.Shuffle(len(locations), func(i, j int) {
+		locations[i], locations[j] = locations[j], locations[i]
+	})
+
+	getLocation := func() (string, error) {
+		if len(locations) == 0 {
+			return "", errors.New("something went wrong")
+		}
+		return locations[0], nil
+	}
+	location, err := getLocation()
+
+	// Package 200 are the fast provisioning servers (40-60min provsion)
+	packageId := 200
+	hostname := strings.Split(uuid.New().String(), "-")[0]
+	domain := pool + ".ofcir"
+
+	key, err := p.ensureSSHKey(p.config.Sshkey)
+	if err != nil {
+		return "", err
+	}
+
+	server := []datatypes.Hardware{
+		{
+			Hostname: sl.String(hostname),
+			Domain:   sl.String(domain),
+		},
+	}
+
+	// Get a list of available keys with
+	// ibmcloud sl order item-list  BARE_METAL_SERVER
+	required_items := []string{
+		"REBOOT_KVM_OVER_IP",
+		"UNLIMITED_SSL_VPN_USERS_1_PPTP_VPN_USER_PER_ACCOUNT",
+		"OS_CENTOS_STREAM_8_X_64_BIT",
+		"BANDWIDTH_0_GB_2",
+		"1_GBPS_PUBLIC_PRIVATE_NETWORK_UPLINKS",
+		"1_IP_ADDRESS",
+	}
+	// Build a skeleton SoftLayer_Product_Item_Price objects.
+	prices := p.getItemPriceList(packageId, required_items)
+
+	// We need to find the ID of the preset with the required name
+	packageService := services.GetProductPackageService(p.client)
+
+	// ibmcloud sl order preset-list  --prices BARE_METAL_SERVER
+	// e.g. presetname == 1U_4210S_384GB_2X4TB_RAID_1
+	// 1U_4210S_384GB_2X4TB_RAID_1, id = 1278
+	presets, err := packageService.Id(packageId).GetActivePresets()
+	presetId := -1
+	for _, preset := range presets {
+		if *preset.KeyName == presetname {
+			presetId = *preset.Id
+			break
+		}
+	}
+	if presetId == -1 {
+		return "", errors.New("Can't find preset with name " + presetname)
+	}
+
+	orderkeys := []datatypes.Container_Product_Order_SshKeys{
+		{
+			SshKeyIds: []int{*key},
+		},
+	}
+
+	// Build a container_Product_Order object.
+	orderTemplate := datatypes.Container_Product_Order{
+		Quantity:         sl.Int(1),
+		Location:         sl.String(location),
+		PackageId:        sl.Int(packageId),
+		PresetId:         sl.Int(presetId),
+		Prices:           prices,
+		Hardware:         server,
+		UseHourlyPricing: sl.Bool(true),
+		SshKeys:          orderkeys,
+	}
+
+	// Get SoftLayer_Product_Order service.
+	service := services.GetProductOrderService(p.client)
+
+	//result, err := service.VerifyOrder(&orderTemplate)
+	_, err = service.PlaceOrder(&orderTemplate, sl.Bool(false))
+	if err != nil {
+		//if strings.Contains(err.Error(), "SoftLayer_Exception_Order_HardwareUnavailable") {
+		return "", err
+	}
+	return hostname, nil
+}
+
+func (p *ibmcloudProvider) Acquire(poolSize int, poolName string, poolType string) (Resource, error) {
+
+	res := Resource{}
+	id, err := p.serverCreate(poolName, p.config.Preset)
+	if err != nil {
+		return res, err
+	}
+
+	// Node don't appear in ibmcloud immediatly, Should we wait for the node to appear?
+	// would prevent a bunch of node not found errors in the logs from acquirecomplete
```

Comment:
The `Acquire` and `AcquireCompleted` were meant to handle this kind of async behaviors. Making the `Acquire` a blocking call will impact negatively the operator behavior (`MaxConcurrentReconciles` is set to `)

Reply:
Yup, thats why I didn't add a wait for the node to appear in the node lists, as a result if we run this code as is
acquire complete will complain about "node not found" the first few times it is run.